### PR TITLE
feat(redis): Allow distinction between exceptions

### LINKED ISSUE #90: More descriptive errors
There are two separate places in aioredlock/redis.py where a LockError is raised with the exact same message - 'Can not get lock'. There is no information about the underlying cause of these errors - the preceding error gets swallowed. 

Furthermore, it seems like a LockError can be caused by two very different situations: either the lock is already acquired by something (perfectly normal, not something you would be concerned about - you would probably just want to swallow these errors) OR a real error happened when connecting to redis (this is something that you would potentially NOT want to swallow and instead would want to debug). It would be very useful if these two situations threw different errors if possible.

Also I think there is a typo in aioredlock/redis.py@unset_lock: there is a LockError raised where the message should be 'Can not unset lock' but instead it reads 'Can not set lock'

I think changes to error reporting could make debugging when using this library easier

### PR DESCRIPTION
**Pull request by bravier**
This commit implements 2 things to enhance errors management.

1. It implements exceptions chaining in `Redis.set_lock`,
   `Redis.get_lock_ttl` and `Redis.unset_lock` so calling code of these
   methods is aware of the original exceptions that prevented the operation
   to complete.
2. It creates new exception types derived from `LockError` to distinguish
   between "normal operation" errors (if a lock is already acquired there's
   no need to raise an exception) and "real errors" like if something
   prevents us to connect to Redis

Closes #90.

### CODE REVIEW

File: aioredlock/redis.py [hunk 1]
```diff
@@ -400,7 +402,8 @@ async def unset_lock(self, resource, lock_identifier):
                        resource, successful_removes, len(self.instances), elapsed_time)
 
         if not unlocked:
-            raise LockError('Can not release the lock')
+            errors = [e for e in successes if isinstance(e, Exception)] or [None]
+            raise LockError('Can not release the lock') from errors[0]
```
Comment (gtmanfred) [inline aioredlock/redis.py:406]: Unfortunately, raising from [None] is going to raise an undesired effect I think
```
>>> raise Exception('this') from [None]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: exception causes must derive from BaseException
```
Shouldn't there be an exception tin errors if it was unable to lock?  Why do we need the or statement.

Comment (gtmanfred) [reply aioredlock/redis.py:406]: If this is not the case, can we write a test that enforces this situation, where the [None] would be used?

File: tests/ut/test_redis.py [hunk 2]
```diff
@@ -464,8 +464,10 @@ async def test_three_instances_combination(
         if success:
             await method('resource', 'lock_id')
         else:
-            with pytest.raises(LockError):
+            with pytest.raises(LockError) as exc_info:
                 await method('resource', 'lock_id')
+            assert (hasattr(exc_info.value, '__cause__') and
```
[the remark this answers is not in the document]
Comment (bravier) [reply tests/ut/test_redis.py:469]: Sure! Updated with:
```diff
--- a/tests/ut/test_redis.py
+++ b/tests/ut/test_redis.py
@@ -466,8 +466,8 @@ class TestRedis:
         else:
             with pytest.raises(LockError) as exc_info:
                 await method('resource', 'lock_id')
-            assert (hasattr(exc_info.value, '__cause__') and
-                    isinstance(exc_info.value.__cause__, (type(None), Exception)))
+            assert hasattr(exc_info.value, '__cause__')
+            assert isinstance(exc_info.value.__cause__, BaseException)
 
         script_sha1 = getattr(redis.instances[0],
                               '%s_script_sha1' % method_name)
```

File: aioredlock/redis.py (hunk 1 above)
Comment (bravier) [reply aioredlock/redis.py:406]: Hi @gtmanfred !

> Unfortunately, raising from [None] is going to raise an undesired effect I think

The code won't raise from `[None]` but from `None` (because of `errors[0]` and `[None][0] = None`).

> Shouldn't there be an exception tin errors if it was unable to lock? Why do we need the or statement.

In the previous version, I was ignoring `asyncio.CancelledError` (the only error that don't inherit from `Exception` but from `BaseException`) and stopping the chain (`raise ... from None`) for them.
But that was a bad idea and I've updated the code to chain `asyncio.CancelledError`s as well with:
```diff
--- a/aioredlock/redis.py
+++ b/aioredlock/redis.py
@@ -342,7 +342,7 @@ class Redis:
                        resource, successful_sets, len(self.instances), elapsed_time)
 
         if not locked:
-            errors = [e for e in successes if isinstance(e, Exception)] or [None]
+            errors = [e for e in successes if isinstance(e, BaseException)]
             raise LockError('Can not acquire the lock "%s"' % resource) from errors[0]
 
         return elapsed_time
@@ -372,7 +372,7 @@ class Redis:
                        resource, len(successful_list), len(self.instances), elapsed_time)
 
         if not success:
-            errors = [e for e in successes if isinstance(e, Exception)] or [None]
+            errors = [e for e in successes if isinstance(e, BaseException)]
             raise LockError('Could not fetch the TTL for lock "%s"' % resource) from errors[0]
 
         return successful_list[0]
```
But there's a catch un `unset_lock` because it can be called with `ensure_future` if we get a `CancelledError` in `set_lock`:
https://github.com/joanvila/aioredlock/blob/49dc1aff64c50bc59c17735d09a2922e03171f1c/aioredlock/algorithm.py#L96-L101

In this case, `if not unlocked:` is `True` (because `successes` list is empty) and if don't have a fallback value (`or [None]`) for exception chaining then we get a list index error.
What do you think?
(PS: I'm not an expert and I may have missed something so don't hesitate to tell!)

**Comment by bravier** [comment]
Just pushed a new version to resolve conflicts.

File: aioredlock/redis.py (hunk 1 above)
Comment (gtmanfred) [reply aioredlock/redis.py:406]: Is it possible to write a test for this edge case?

Comment (gtmanfred) [reply aioredlock/redis.py:406]: Is it possible to write a test for this edge case?

**Review by gtmanfred — APPROVED** [review]
I just had one question, and then we can get this merged and released. Thanks for all the work on this one. I think it is going to turn out really good.

**Comment by gtmanfred** [comment]
```
>>> raise Exception() from Exception('hi')
Exception: hi

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
Exception
```
Nice, i always wondered how these exceptions got chained, this is going to be really helpful when troubleshooting some issues.

File: aioredlock/redis.py (hunk 1 above)
Comment (bravier) [reply aioredlock/redis.py:406]: I would have love to but when coming back to this issue I finally understood what was happening here..

As I showed in the updated examples, I am using this code structure with aioredlock:
```py
try:
  lock = await lock_manager.lock()
except LockError as e:
  ... log error or raise
else:
  ... if no exception, run the code protected by the lock
  await lock_manager.unlock(lock)
finally:
  await lock_manager.destroy()
```
In this case, the `finally` block will be executed whether or not there was an error when acquiring the lock.
But if we get an exception in `set_lock` then the code in `algorithm.py` will catch it and schedule a task to cleanup later:
https://github.com/joanvila/aioredlock/blob/49dc1aff64c50bc59c17735d09a2922e03171f1c/aioredlock/algorithm.py#L96-L101
It conflicts with the `destroy` we do ourselves which itself calls `unset_lock` and destroys our `Instance` objects.
When the scheduled task is executed, there is no instances to clean anymore as they were destroyed and with the conditions in `unset_lock` we end up here:
https://github.com/joanvila/aioredlock/blob/49dc1aff64c50bc59c17735d09a2922e03171f1c/aioredlock/redis.py#L403

That's why I had to protect the `raise LockError(...) from [None][0]` in my previous comment.
When not using `finally` in my code then no problem anymore and I can remove the default value to raise from.

Unfortunately this reveals an issue with the current Aioredlock code that need to be fixed in another PR: if there is an exception on one of the instance when setting a lock then resources acquired on other instances are never cleaned up. Not cool.

Anyway, I've pushed a new version of the PR.

Comment (gtmanfred) [reply aioredlock/redis.py:406]: Perfect, thanks for the detail.  This looks good now. Lets get it merged, and then we can address the other issue.
